Add Instrumentation event

kylefox opened the pull request:
I'd like a bit more control over how we log spam detection, so I added a [custom event](https://guides.rubyonrails.org/active_support_instrumentation.html#creating-custom-events) using `ActiveSupport::Notifications`.

The `'invisible_captcha.spam_detected'` event fires when spam is detected, allowing the user (i.e. developers) to handle custom logging to a database, Rollbar, etc:

```ruby
ActiveSupport::Notifications.subscribe 'invisible_captcha.spam_detected' do |name, started, finished, unique_id, data|
  #
  # `data` is a hash with the following attributes:
  #
  # {
  #   message: "Invisible Captcha honeypot param 'zclwijk-uxpyqtf' was present.",
  #   remote_ip: "127.0.0.1",
  #   user_agent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.97 Safari/537.36",
  #   controller: "users",
  #   action:"create",
  #   url:"https://app.lvh.me:5001/users",
  #   params: <ActionController::Parameters {...} >
  # }

  # Log all this context to Rollbar:
  Rollbar.warn(data[:message], data)

  # Record the blocked request in your database:
  SpamRequest.create(data.except(:params))
end
```

~I included `params` because it will be helpful in identifying false positives: I can look at the rest of the submitted form data to see if it was a legitimate request or indeed spam. There are some implications to this — developers will need to ensure they are not accidentally logging sensitive parameters. But this seems like a reasonable tradeoff, given this is "advanced usage" anyway.~

☝️The event is now passed filtered (safe) params https://github.com/markets/invisible_captcha/pull/62/commits/e43fa7d97764d032d4f17a5fe6fc4acb473f4a19

If you generally like the idea, I'll add tests and a guard condition for older versions of Rails that don't have `ActiveSupport::Notifications`.

Let me know what you think!

- [x] Add tests
- [x] Update README
- [x] ~Add guard clause for older versions of Rails~

markets commented:
@kylefox seems really useful to have this "enabled" by default :ok_hand: +1

I'm actually doing something similar using the controller callbacks:

```ruby
invisible_captcha only: [:create], on_spam: :handle_spam
```

But having a default instrumentation via `ActiveSupport::Notifications` seems useful for everybody. 

I think we probably don't need a guard for older Rails, if I'm not wrong, it was introduced in Rails 3 (our minor supported version). 

Thanks!

kylefox commented:
@markets I added tests (https://github.com/markets/invisible_captcha/pull/62/commits/3bf08d6c52b6f0ee590655013cafe4fccd0fea89) and updated the README (https://github.com/markets/invisible_captcha/pull/62/commits/78488785b82019f100f845a346773473edd5beeb) but CI is failing for Rails 4.2 and 3.2.

I've run into some issues with `bundle exec appraisal install` for Rails 4.2 (and presumably 3.2):

```
❯ bundle exec appraisal install
>> bundle check --gemfile='/Users/kyle/Code/invisible_captcha/gemfiles/rails_6.0.gemfile' || bundle install --gemfile='/Users/kyle/Code/invisible_captcha/gemfiles/rails_6.0.gemfile' --retry 1
The Gemfile's dependencies are satisfied
>> bundle check --gemfile='/Users/kyle/Code/invisible_captcha/gemfiles/rails_5.2.gemfile' || bundle install --gemfile='/Users/kyle/Code/invisible_captcha/gemfiles/rails_5.2.gemfile' --retry 1
The Gemfile's dependencies are satisfied
>> bundle check --gemfile='/Users/kyle/Code/invisible_captcha/gemfiles/rails_5.1.gemfile' || bundle install --gemfile='/Users/kyle/Code/invisible_captcha/gemfiles/rails_5.1.gemfile' --retry 1
The Gemfile's dependencies are satisfied
>> bundle check --gemfile='/Users/kyle/Code/invisible_captcha/gemfiles/rails_5.0.gemfile' || bundle install --gemfile='/Users/kyle/Code/invisible_captcha/gemfiles/rails_5.0.gemfile' --retry 1
The Gemfile's dependencies are satisfied
>> bundle check --gemfile='/Users/kyle/Code/invisible_captcha/gemfiles/rails_4.2.gemfile' || bundle install --gemfile='/Users/kyle/Code/invisible_captcha/gemfiles/rails_4.2.gemfile' --retry 1
Bundler can't satisfy your Gemfile's dependencies.
Install missing gems with `bundle install`.
Fetching gem metadata from https://rubygems.org/.............
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Bundler could not find compatible versions for gem "bundler":
  In rails_4.2.gemfile:
    bundler (~> 1.17)

  Current Bundler version:
    bundler (2.0.2)
This Gemfile requires a different version of Bundler.
Perhaps you need to update Bundler by running `gem install bundler`?

Could not find gem 'bundler (~> 1.17)' in any of the relevant sources:
  the local ruby installation
```

I'm able to appraise Rails > 4 just fine (ex: `bundle exec appraisal rails-5.2 rspec`) but I'm just not familiar enough with Appraisal to figure out Rails < 4 😕Any suggestions?

markets commented:
Hi @kylefox :wave: 

The problem seems similar to what we need in CI (Rails 4 doesn't support bundler 2):

https://github.com/markets/invisible_captcha/blob/87ffa1a973095718857428d763fd50557d6a6c35/.travis.yml#L18-L22

Checking the failed builds in Travis (ex: https://travis-ci.org/markets/invisible_captcha/jobs/609341878), seems related to some Rails internal changes (key vs string)... not sure how we can fix this, maybe we can relax a bit the spec?

kylefox commented:
That seems to be fixed, but there's still an issue with the `:url` attribute in Rails 3 and 4.

https://travis-ci.org/markets/invisible_captcha/jobs/609746727#L359

It looks like 3 & 4 are including query parameters in the URL, which is odd because the request is supposed to be a `post` — the params should be in the request body, not the URL 🤔It may be something related to `switchable_post`

However, I'm afraid I can't really justify spending much more time on this 😕 I still can't get Rails 3 & 4 specs working locally, making this extremely difficult to troubleshoot. And I'm already using my fork in production...

If you have any more clues about why this spec is failing, I can take another crack at it. Otherwise we'll need to remove the `with` matcher from the spec, which isn't ideal.

markets commented:
> still an issue with the :url attribute in Rails 3 and 4

I see... weird 🤔 => "the params should be in the request body" that's it

> may be something related to switchable_post

Code is just a regular post passing the params: 

https://github.com/markets/invisible_captcha/blob/87ffa1a973095718857428d763fd50557d6a6c35/spec/controllers_spec.rb#L10

So I don't think so.

> I'm afraid I can't really justify spending much more time on this

Yep, agree, the code and tests for active Rails versions are totally fine :ok_hand: Maybe we can add a conditional here... Like checking a different hash for Rails < 5, not ideal but at least we keep `with` matcher. What do you think? I'd really like to merge this. 

Thanks a lot for working on this!

kylefox commented:
@markets Done, CI seems to be passing now 👍

markets commented:
@kylefox released v1.0.0 :rocket: (contains this feature)
